app/controllers/concerns/foreman/controller/smart_proxy_auth.rb

```diff
@@ -94,11 +94,26 @@ def auth_smart_proxy(proxies = SmartProxy.all, require_cert = true)
     hosts = Hash[proxies.map { |p| [URI.parse(p.url).host, p] }]
     allowed_hosts = hosts.keys.push(*Setting[:trusted_puppetmaster_hosts])
     logger.debug { ("Verifying request from #{request_hosts.inspect} against #{allowed_hosts.inspect}") }
-    unless host = allowed_hosts.detect { |p| request_hosts.include? p }
+
+    if (host = detect_matching_host(allowed_hosts, request_hosts))
+      @detected_proxy = hosts[host] if host
+      true
+    else
       logger.warn "No smart proxy server found on #{request_hosts.inspect} and is not in trusted_puppetmaster_hosts"
-      return false
+      false
     end
-    @detected_proxy = hosts[host] if host
-    true
+  end
+
+  def detect_matching_host(allowed_hosts, request_hosts)
+    allowed_hosts.product(request_hosts).each do |allowed, request|
+      if request.starts_with?('*')
+        rex = /\A#{Regexp.escape(request).sub('\\*', '.*')}\Z/
+        return allowed if allowed =~ rex
+      else
+        return allowed if allowed == request
+      end
+    end
+
+    nil
```

Comment:
> .each returns nil anyway

I don't think that's true, `.each` returns the object on which it has been run.

short test confirms
```ruby
[1,2,].each { |x| nil }
=> [1, 2]
```

 Even with map instead of each there, the result would be array of `nil`s. Removing the WoC